### PR TITLE
Plot epochs.events by default

### LINKED ISSUE #11439: Should `mne.Epochs.plot()` plot events and IDs by default?
An `mne.Epochs` instance has associated `events` and `event_id` attributes. These correspond to the events and IDs set during epoch creation.

When plotting existing epochs with `epochs.plot()`, no events are drawn. If events should be included, they need to be passed explicitly, e.g. `epochs.plot(events=epochs.events, event_id=epochs.event_id)`. To me, this is rather cumbersome, since I expect the epoch labels to be included by default. However, maybe there is a good reason for this behavior – if not, I'd be happy to contribute a PR to include events by default.

### PR DESCRIPTION
**Pull request by cbrnr**
Fixes #11439, which means that `epochs.events` are now automatically plotted (together with `epochs.event_id`).

I suggest the following changes to the `events` and `event_id` parameters: First, automatic plotting should be enabled when `events=True`. Second, this should be the new default. That way, users can still choose to plot no events with `events=None` (the previous default).

### CODE REVIEW

**Comment by cbrnr** [comment]
The pyvista issues are probably not related, but who knows. I'll be reiterating on the PR anyway, so maybe it goes away on the next push.

**Comment by cbrnr** [comment]
Before I flesh this out more, is everyone on board with this change?

- The new default is `events=True` and `event_id=True`, which means use `epochs.events` and `epochs.event_id`.
- `events=None` and `event_id=None` still works and plots no events, but this is not the default anymore.
- Custom events are also still possible by passing `events` and `event_id`.

It's maybe only a bit weird using `True`, but `None` was already taken and I think implies "no events". It's not clear what should happen when users pass `False` though. Maybe using a string (e.g. "auto") would be better?

**Comment by agramfort** [comment]
+1 on my side

**Comment by drammock** [comment]
> It's maybe only a bit weird using `True`, but `None` was already taken and I think implies "no events". It's not clear what should happen when users pass `False` though. Maybe using a string (e.g. `"auto"`) would be better?

There are only 3 behaviors we need to support:

1. don't plot events
2. plot events automatically
3. plot custom events

We can support that with `False`, `True`, `ndarray` (or `False`, `True`, `dict` for the event_id).  I recommend doing that and (silently) continuing to accept `None` as a synonym for `False`.  I've done that elsewhere in the codebase so it's not unprecedented.  Note that since the current default is `None` we'll need a deprecation cycle to change the default to `True`, so you'll need the `if events is None: warn("whatever")` conditional anyway... when the deprecation cycle ends you just change that to `if events is None: events = False`.

**Comment by cbrnr** [comment]
I still have two questions:

1. The deprecation cycle means that the default does not (can not) change right now, correct?
2. I like the suggested values for `events`, because they directly map to plotting no events (`events=False`), integrated events (`events=True`), and custom events (`events=events`). However, this is not so clear for `event_id`. Currently, `event_id=None` means plotting events using `epochs.event_id` (I think?). It looks like this argument is combined with `epochs.event_id` and it doesn't replace it. Therefore, the default should already do the right thing. But I'm not really sure about that, because I didn't find it in the documentation.

**Comment by drammock** [comment]
> The deprecation cycle means that the default behavior does not (cannot?) change right now, correct?

default behavior does not change, but should raise a warning. So for now:

```py
if events is None:
    warn('In version XXX the default will change from None to True, set explicitly to False to avoid this warning.', FutureWarning)
    events = False
``` 

> To silence the `DeprecationWarning` in our tests, I've set `events=False`. I could also capture the warning – what do you prefer? After the deprecation cycle, we should remove `events=False` again, so maybe capturing is preferable?

If it were me, I'd leave it as `events=False` and add a descriptive `# TODO` comment about what to change and when.  But the `catch_warnings` approach is fine with me too.

> I like the suggested values for `events`, because they directly map to plotting no events (`events=False`), integrated events (`events=True`), and custom events (`events=events`). However, this is not so clear for `event_id`. Ideally, if `events=True` then `event_id` should be set to `epochs.event_id`. It doesn't make sense to set this separately. Any idea how to best do this?

The current docstring is actually quite confusing on this point. To me it reads as if `epochs.event_id` is already getting used automatically (it isn't) unless the user passes a separate dict, in which case the user-passed dict will override `epochs.event_id` when there are key clashes (in fact, if I pass `event_id=dict(buttonpress=32)` with the sample data, then *only the buttonpress events* get text labels, the other events are still just integer labels, i.e., `epochs.event_id` is not getting used there either).

I think the behavior I would want/expect is:

1. `events=False` don't show anything
2. `events=True, event_id=False` show integer labels
3. `events=True, event_id=True` show text labels from `epochs.event_id`
4. `events=True, event_id=dict` use both dicts and prefer user-passed dict for key clashes
5. `events=array, event_id=False` show integer labels
6. `events=array, event_id=True` show text labels from `epochs.event_id`
7. `events=array, event_id=dict` use both dicts and prefer user-passed dict for key clashes

for cases 6 and 7, also *warn (or perhaps info?) if there are any missing labels* (this is possible because `array` may contain event integers that weren't used during epoching). The backward-compatible case is `event_id=False` (despite what the docs imply!) so the same pattern of `if x is None: warn(); x = False` should still work here.

LMK if I've gotten something wrong here or if any of that didn't make sense.

**Comment by cbrnr** [comment]
Thanks @drammock, this makes perfect sense!

One last question regarding the deprecation cycle (I think I'll go with TODO comments): We issue a `FutureWarning` starting with the next release, i.e. v1.4, right? And then we deprecate and change the (default) behavior in the version after that, so v1.5 – correct?

**Comment by cbrnr** [comment]
Also, what if you really only want to label events that you pass as a dict? Then it doesn't make sense to combine it with an existing `epochs.event_id` dict. I think we should never combine dicts (which is how it works now I think).

**Comment by drammock** [comment]
> Also, what if you really only want to label events that you pass as a dict?

you can get that behavior by populating your custom dict with entries like `"1": 1` or whatever.

> I think we should never combine dicts (which is how it works now I think).

I'd be OK with updating items 4 and 7 from the above list like this:

4. `events=True, event_id=dict` show text labels from user-passed dict
7. `events=array, event_id=dict` show text labels from user-passed dict

I think if that's the plan then there should definitely be an `info` message saying something like "some keys missing, using integer IDs for events 1, 2, and 32" or whatever.

**Comment by cbrnr** [comment]
I'm sorry I keep coming back, but I find this really confusing, so I want to make sure that it works as expected.

When creating epochs with `mne.Epochs()`, the `event_id` dict _selects_ events that are used to create epochs. This means that if there are additional types in `events` that are not in the `event_id` dict, they will be ignored and will not be used to create epochs.

This is different from what we've discussed for `epochs.plot()`, where `event_id` only influences which events are plotted with labels instead of integers. If some types are missing in `event_id`, they will _still_ be plotted (but only with their integer event type).

Do we want this behavior? Or do we want to be consistent with `mne.Epochs()`?

To be honest, all I wanted was to plot the events and event types already contained in the `Epochs` object (ideally by default), so we don't have to change anything else here. I am not really that familiar with the epoching workflow in MNE, so maybe I am overthinking or not understanding it. I thought that plotting events (the ones already available in the `Epochs` object) or plotting no events would be sufficient, but now @drammock listed seven combinations of two parameters, which to me indicates that we should maybe simplify this a bit.

**Comment by agramfort** [comment]
I have not followed closely and I am bit confused. My default idea here would be to plot whatever is available in epochs.event_id which indeeds filters out the events not present in any value of events. How about just having a bool paramter in epochs.plot to display or not the event?

my 2c

**Comment by drammock** [comment]
> When creating epochs with `mne.Epochs()`, the `event_id` dict _selects_ events that are used to create epochs. This means that if there are additional types in `events` that are not in the `event_id` dict, they will be ignored and will not be used to create epochs.

Correct.

> This is different from what we've discussed for `epochs.plot()`, where `event_id` only influences which events are plotted with labels instead of integers. If some types are missing in `event_id`, they will _still_ be plotted (but only with their integer event type). Do we want this behavior? Or do we want to be consistent with `mne.Epochs()`?

Instead of listing the parameter combinations, let's list the by the desired behaviors:

1. **don't plot events**
    a. `events=False, event_id=<ignored>`
2. **plot the events that were used to create the epochs.** These will always fall at `t=0` (which incidentally is I think why it wasn't felt necessary to plot them by default). The exception is when there are overlapping epochs; events might show up at `t!=0` in neighboring epochs.
    a. as integers: `events=True, event_id=False`
    b. as default text: `events=True, event_id=True` (*the new default*)
    c. as custom text:  `events=True, event_id=<dict>` (must decide what to do with events in `epochs.events` but not in `<dict>`)
3. **plot custom events**
    a. as integers: `events=<array>, event_id=False`
    b. as default text: `events=<array>, event_id=True` (must decide what to do with events in `<array>` but not in `epochs.event_id`)
    c. as custom text: `events=<array>, event_id=<dict>` (must decide what to do with events in `<array>` but not in `<dict>`)

There are still 7 options. if you want to grant user control over the "must decide..." parts then there will be even more options.  Please tell us how you think it should be simplified @cbrnr (i.e., which options you want to disallow)

**Comment by cbrnr** [comment]
I am currently reading the [Applied ERP Analysis book](https://socialsci.libretexts.org/Bookshelves/Psychology/Book%3A_Applied_Event-Related_Potential_Data_Analysis_(Luck)) by Steven Luck, and I would like to translate it to MNE commands. One key feature of their ERPLAB toolbox is that they are creating epochs with events of interest while also _keeping_ other (unused) event types.

For example, let's say there are four event types in the raw file:

- 1: frequent
- 2: rare
- 3: correct response
- 4: incorrect response

Say we are interested in frequent vs. rare ERPs, so we use event types 1 and 2 to create epochs. However, we would like to keep only those epochs that are followed by a correct response.

In MNE, I cannot visualize this workflow, because I need to create epochs with all four event types, which I don't want. Of course I could still do it and then pick only the desired event types, but there is no way to create an epochs plot consisting of event types 1 and 2 only, but also showing event types 3 and 4 (if they occur within the epoch window).

What does this have to do with plotting and this PR? Well, I'm not entirely sure. But I think that if it was possible to create epochs with selected event types and still keep the other types, we would probably not need `event_id` at all. Because all event types (and labels) would be included in the epochs object, there would be no need to provide custom events. Maybe the only option could be to show labels or integer codes.

Which would leave only the following options:

1. Don't plot events: `events=False`
2. Plot events: `events=True`

All other use cases could be moved to the epochs creation stage.

So to wrap up, it is entirely possible that I don't know what I'm doing 😄 (I don't have much or even any experience with working with ERPs in MNE)! All I wanted to suggest is to plot `epochs.events` (with `epochs.event_id`) by default, which occured to me as a sensible addition, but now we are discussing much more.

**Comment by agramfort** [comment]
can you see if this example
https://mne.tools/mne-bids-pipeline/1.1/examples/ERP_CORE.html#configuration
does what you suggest?

the important lines are:

conditions = { "related": '`first_stimulus/target` == "related" and '
'first_response
== "correct"', "unrelated": '`first_stimulus/target` == "unrelated"
and ' 'first_response
== "correct"', }

**Comment by cbrnr** [comment]
@agramfort pretty much, yes, but it uses MNE-BIDS-Pipeline. Does this way of extracting conditions also work in MNE-Python?

**Comment by agramfort** [comment]
it mostly uses
https://mne.tools/stable/generated/mne.epochs.make_metadata.html

**Comment by drammock** [comment]
> In MNE, I cannot visualize this workflow, because I need to create epochs with all four event types, which I don't want. Of course I could still do it and then pick only the desired event types, but there is no way to create an epochs plot consisting of event types 1 and 2 only, but also showing event types 3 and 4 (if they occur within the epoch window).

This is incorrect. Please look again at https://github.com/mne-tools/mne-python/issues/11439#issuecomment-1414450126

**Comment by cbrnr** [comment]
I saw your MWE @drammock, but I thought that you needed to include the extra epochs. Thanks for clarifying. My MWE still doesn't show the extra events, but this seems to be a user error, so I will move that question to our forum.

Coming back to this issue, I can see now that the separation between `epochs.events`, `epochs.event_id`, as well as custom `events` and `events_id` for plotting are necessary in the MNE workflow, and the ERPLAB workflow cannot be translated 1:1. The only thing I'd still change is to plot the included events by default, nothing more.

**Comment by drammock** [comment]
@cbrnr any interest in coming back to this one? It seemed to me like it was pretty close to ready (now has a merge conflict but nothing too major I don't think?)

**Comment by cbrnr** [comment]
Yes, I will need to rebase and test this to remind myself of what I did. I can't do it right now because of end of semester workload, so if you want it earlier please feel free to make the last finishing touches (if you have time).

**Comment by drammock** [comment]
OK I rebased and force-pushed; the rebase was needed just because of switching to use Black (no complicated merge conflicts involved). along the way I found the bug that was causing the weird behavior in https://mne.discourse.group/t/6334/

File: mne/viz/epochs.py [hunk 1]
```diff
@@ -946,12 +945,27 @@ def plot_epochs(
     boundary_times = np.arange(len(epochs) + 1) * len(epochs.times) / sfreq
 
     # events
-    if events is not None:
+    if events is None:
+        warn(
+            "The current default events=None is deprecated and will change to "
+            "events=True in MNE 1.5. Set events=False to suppress this "
```
Comment (drammock) [inline mne/viz/epochs.py:951]:
```suggestion
            "events=True in MNE 1.6. Set events=False to suppress this "
```

File: mne/viz/tests/test_epochs.py [hunk 2]
```diff
@@ -38,63 +41,65 @@ def test_plot_epochs_basic(epochs, epochs_full, noise_cov_io, capsys, browser_ba
...
+        epochs.plot(**ev, epoch_colors=[["r"]])
     # also test event_color
-    epochs.plot(event_color="b")
+    epochs.plot(**ev, event_color="b")
 
 
 def test_plot_epochs_scale_bar(epochs, browser_backend):
     """Test scale bar for epochs."""
-    fig = epochs.plot()
+    fig = epochs.plot(events=False)
```
Comment (drammock) [inline mne/viz/tests/test_epochs.py:102]:
```suggestion
    fig = epochs.plot(**ev)
```

File: mne/viz/tests/test_epochs.py [hunk 3]
```diff
@@ -147,7 +152,7 @@ def test_plot_epochs_clicks(epochs, epochs_full, capsys, browser_backend):
     fig._close_event()  # XXX workaround, MPL Agg doesn't trigger close event
     assert len(epochs_full) == 6
     # test rightclick → image plot
-    fig = epochs_full.plot()
+    fig = epochs_full.plot(events=False)
```
Comment (drammock) [inline mne/viz/tests/test_epochs.py:155]:
```suggestion
    fig = epochs_full.plot(**ev)
```

File: mne/viz/tests/test_epochs.py [hunk 4]
```diff
@@ -227,7 +237,7 @@ def test_plot_epochs_nodata(browser_backend):
     info = create_info(2, 1000.0, "stim")
     epochs = EpochsArray(data, info)
     with pytest.raises(ValueError, match="consider passing picks explicitly"):
-        epochs.plot()
+        epochs.plot(events=False)
```
Comment (drammock) [inline mne/viz/tests/test_epochs.py:240]:
```suggestion
        epochs.plot(**ev)
```

File: mne/viz/tests/test_epochs.py [hunk 5]
```diff
@@ -413,11 +423,11 @@ def test_plot_epochs_ctf(raw_ctf, browser_backend):
     )
     evts = make_fixed_length_events(raw_ctf)
     epochs = Epochs(raw_ctf, evts, preload=True)
-    epochs.plot()
+    epochs.plot(events=False)
```
Comment (drammock) [inline mne/viz/tests/test_epochs.py:426]:
```suggestion
    epochs.plot(**ev)
```

**Review by drammock — COMMENTED** [review]
some little things I missed. waiting for current CI run to finish in case there are more failures to add to this batch.

**Comment by cbrnr** [comment]
Looks good! I checked with the example I tried in the forum post, and it works now!

One caveat of showing multiple events per epoch: now it's hard to identify events used as time-locking events for creating epochs. We could e.g. show the corresponding events in bold – but maybe this should be tackled in another PR.

**Comment by drammock** [comment]
> One caveat of showing multiple events per epoch: now it's hard to identify events used as time-locking events for creating epochs. We could e.g. show the corresponding events in bold

I took a quick look and it's not trivial so I opened #11751 to track that request.

This one is thus ready for review/merge

**Comment by larsoner** [comment]
LGTM just needs a `latest.inc` entry in API section, feel free to add it then mark for merge when green

**Comment by cbrnr** [comment]
Thanks @drammock!

**Comment by larsoner** [comment]
Looks like this caused some future warnings and at least one repeated kwarg error

https://app.circleci.com/pipelines/github/mne-tools/mne-python/19407/workflows/351149c0-a8b7-48e9-b87a-15e299bffbdb/jobs/55471

@drammock can you look?

**Comment by drammock** [comment]
Dang. Yes I'll look ASAP

**Comment by larsoner** [comment]
I might have a minute here, I'll take a look

**Pull request merged**